### PR TITLE
bpo-15913 : Add PyBuffer_SizeFromFormat()

### PR DESCRIPTION
**Pull request by nanjekyejoannah**
Added implementation for PyBuffer_SizeFromFormat().

https://bugs.python.org/issue15913

### CODE REVIEW

**Comment by nanjekyejoannah** [comment]
Requesting review @skrah @pitrou @vstinner

**Comment by pablogsal** [comment]
I think this needs tests. Implementation aside, calling the current function with `3si 3s 0i` segfaults for me.

**Comment by nanjekyejoannah** [comment]
@pablogsal , thought about that  also but am still wondering where to place this test?

**Comment by pablogsal** [comment]
> @pablogsal , thought about that also but am still wondering where to place this test?

I would say exposed in `Modules/_testcapimodule.c` and then tested with `Lib/test/test_capi.py` but Victor or Antoine would know better regarding this specific API.

File: Doc/c-api/buffer.rst [hunk 1]
```diff
@@ -460,10 +460,11 @@ Buffer-related functions
    :c:func:`PyObject_GetBuffer`.
 
 
-.. c:function:: Py_ssize_t PyBuffer_SizeFromFormat(const char *)
+.. c:function:: Py_ssize_t PyBuffer_SizeFromFormat(const char *format)
 
    Return the implied :c:data:`~Py_buffer.itemsize` from :c:data:`~Py_buffer.format`.
-   This function is not yet implemented.
+
+   .. versionchanged:: 3.9
```
Comment (vstinner) [inline Doc/c-api/buffer.rst]: the function is new in 3.9 no? In that case, it should be "versionadded:: 3.9".

File: Objects/abstract.c [hunk 2]
```diff
@@ -495,6 +500,30 @@ _Py_add_one_to_index_C(int nd, Py_ssize_t *index, const Py_ssize_t *shape)
     }
 }
 
+int
+PyBuffer_SizeFromFormat(const char *format)
+{
+    PyObject *result;
+    int size;
+
+    structmodule = PyImport_ImportModule("struct");
```
Comment (vstinner) [inline Objects/abstract.c:507]: it's a strong reference: you must Py_DECREF(structmodule) to prevent a reference leak. I suggest to add a "done:" label which does DECREF and replace return NULL with "goto done".

File: Objects/abstract.c [hunk 3]
```diff
@@ -495,6 +500,30 @@ _Py_add_one_to_index_C(int nd, Py_ssize_t *index, const Py_ssize_t *shape)
...
+    if (structmodule == NULL)
+        return NULL;
```
Comment (vstinner) [inline Objects/abstract.c]: wait, NULL is a pointer. The function return type is int: you should return -1 in case of error, no? Same code for others if below.

File: Doc/c-api/buffer.rst [hunk 4]
```diff
@@ -460,10 +460,11 @@ Buffer-related functions
    :c:func:`PyObject_GetBuffer`.
 
 
-.. c:function:: Py_ssize_t PyBuffer_SizeFromFormat(const char *)
+.. c:function:: Py_ssize_t PyBuffer_SizeFromFormat(const char *format)
 
    Return the implied :c:data:`~Py_buffer.itemsize` from :c:data:`~Py_buffer.format`.
```
Comment (vstinner) [inline Doc/c-api/buffer.rst:465]: Please document how errors are reported. I suggest to return -1.

**Review by vstinner — CHANGES_REQUESTED** [review]

**Comment by vstinner** [comment]
Would it be possible to add an unit test? Expose the function in _testcapi and add a test in test_buffer using _testcapi (such test should be marked with ``@cpython_only``, or skipped if _testcapi is None).

**Comment by nanjekyejoannah** [comment]
I have made some changes locally that will address some of the comments @vstinner. I will update the PR when I finish the tests.

**Comment by vstinner** [comment]
> I have made some changes locally that will address some of the comments @vstinner. I will update the PR when I finish the tests.

Ok. Ping me when I should review again ;-)

File: Lib/test/test_buffer.py [hunk 5]
```diff
@@ -4412,6 +4412,24 @@ def test_issue_7385(self):
...
         self.assertRaises(BufferError, memoryview, x)
 
+    # Test PyBuffer_SizeFromFormat()
+    @support.cpython_only
+    def test_pybuffer_size_from_format(self):
+        from ctypes import pythonapi
+        PyBuffer_SizeFromFormat = pythonapi.PyBuffer_SizeFromFormat
+
+        # basic tests
+        self.assertIs(PyBuffer_SizeFromFormat("3si") > 0 , True)
```
Comment (vstinner) [inline Lib/test/test_buffer.py]: I suggest to test a few formats using a loop and compare the result to struct.calcsize(). Something like:

```
for format in ("i", "3s", "0i", ""):
    self.assertEqual(PyBuffer_SizeFromFormat(format.encode()), struct.calcsize(format))
```

You must explicitly encode the Unicode format string to UTF-8 to pass it as a byte string (char*), no?

File: Objects/abstract.c [hunk 6]
```diff
@@ -495,6 +494,47 @@ _Py_add_one_to_index_C(int nd, Py_ssize_t *index, const Py_ssize_t *shape)
...
+
+    fmt = PyBytes_FromString(format);
+    if (fmt == NULL)
+        goto error;
+
+    structobj = PyObject_CallFunctionObjArgs(Struct, fmt, NULL);
+    if (structobj == NULL)
+        goto error;
+
+    itemsize = (intptr_t)structobj;
```
Comment (vstinner) [inline Objects/abstract.c]: I don't think that it works. structobj is a Python object. You should use PyLong_AsSsize_t() to get it as Py_ssize_t.

File: Objects/abstract.c [hunk 7]
```diff
@@ -495,6 +494,47 @@ _Py_add_one_to_index_C(int nd, Py_ssize_t *index, const Py_ssize_t *shape)
     }
 }
 
+int
```
Comment (vstinner) [inline Objects/abstract.c]: The doc uses Py_ssize_t, but you chose int. I would prefer to use Py_ssize_t.

File: Objects/abstract.c [hunk 8]
```diff
@@ -495,6 +494,47 @@ _Py_add_one_to_index_C(int nd, Py_ssize_t *index, const Py_ssize_t *shape)
...
+PyBuffer_SizeFromFormat(const char *format)
+{
+    PyObject *structmodule;
+    PyObject *Struct = NULL;
+    PyObject *structobj = NULL;
+    PyObject *fmt = NULL;
+    int itemsize;
```
Comment (vstinner) [inline Objects/abstract.c]: initialize itemsize to -1, so the error label can be removed, and you can just use "goto done".

File: Objects/abstract.c [hunk 9]
```diff
@@ -495,6 +494,47 @@ _Py_add_one_to_index_C(int nd, Py_ssize_t *index, const Py_ssize_t *shape)
     }
 }
 
+int
+PyBuffer_SizeFromFormat(const char *format)
+{
+    PyObject *structmodule;
+    PyObject *Struct = NULL;
+    PyObject *structobj = NULL;
```
Comment (vstinner) [inline Objects/abstract.c]: I suggest to rename it to "res".

**Review by vstinner — CHANGES_REQUESTED** [review]

File: Lib/test/test_buffer.py [hunk 10]
```diff
@@ -4412,6 +4412,24 @@ def test_issue_7385(self):
         x = ndarray([1,2,3], shape=[3], flags=ND_GETBUF_FAIL)
         self.assertRaises(BufferError, memoryview, x)
 
+    # Test PyBuffer_SizeFromFormat()
+    @support.cpython_only
+    def test_pybuffer_size_from_format(self):
+        from ctypes import pythonapi
+        PyBuffer_SizeFromFormat = pythonapi.PyBuffer_SizeFromFormat
```
Comment (pablogsal) [inline Lib/test/test_buffer.py]: This test should be able to run even if `ctypes` is not compiled. Could you expose this function in `Modules/_testcapimodule.c`?

File: Objects/abstract.c [hunk 11]
```diff
@@ -495,6 +494,47 @@ _Py_add_one_to_index_C(int nd, Py_ssize_t *index, const Py_ssize_t *shape)
...
+    Py_DECREF(structmodule);
+    if (Struct == NULL)
+        goto error;
+
+    fmt = PyBytes_FromString(format);
+    if (fmt == NULL)
+        goto error;
+
+    structobj = PyObject_CallFunctionObjArgs(Struct, fmt, NULL);
+    if (structobj == NULL)
```
Comment (pablogsal) [inline Objects/abstract.c]: All of these conditionals should comply with pep7 (use braces).

File: Objects/abstract.c [hunk 12]
```diff
@@ -495,6 +494,47 @@ _Py_add_one_to_index_C(int nd, Py_ssize_t *index, const Py_ssize_t *shape)
...
+PyBuffer_SizeFromFormat(const char *format)
+{
+    PyObject *structmodule = NULL;
+    PyObject *Struct = NULL;
+    PyObject *res = NULL;
+    PyObject *fmt = NULL;
+    Py_ssize_t itemsize = -1;
+
+    structmodule = PyImport_ImportModule("struct");
+    if (structmodule == NULL)
```
Comment (vstinner) [inline Objects/abstract.c]: Nitpick: PEP 7 requires {...} here.

File: Doc/c-api/buffer.rst [hunk 13]
```diff
@@ -460,10 +460,12 @@ Buffer-related functions
    :c:func:`PyObject_GetBuffer`.
 
 
-.. c:function:: Py_ssize_t PyBuffer_SizeFromFormat(const char *)
+.. c:function:: Py_ssize_t PyBuffer_SizeFromFormat(const char *format)
 
-   Return the implied :c:data:`~Py_buffer.itemsize` from :c:data:`~Py_buffer.format`.
-   This function is not yet implemented.
+   Return the implied :c:data:`~Py_buffer.itemsize` from :c:data:`~Py_buffer.format`. on success, the
+   size of the return *format* and -1 Otherwise.
```
Comment (vstinner) [inline Doc/c-api/buffer.rst]: Sorry, I don't understand the second sentence. I would suggest to replace it with:

Raise an exception and return -1 on error.

File: Objects/abstract.c [hunk 14]
```diff
@@ -495,6 +494,47 @@ _Py_add_one_to_index_C(int nd, Py_ssize_t *index, const Py_ssize_t *shape)
...
+Py_ssize_t
+PyBuffer_SizeFromFormat(const char *format)
+{
+    PyObject *structmodule = NULL;
+    PyObject *Struct = NULL;
+    PyObject *res = NULL;
+    PyObject *fmt = NULL;
+    Py_ssize_t itemsize = -1;
+
+    structmodule = PyImport_ImportModule("struct");
```
Comment (vstinner) [inline Objects/abstract.c:507]: You are leaking one reference to structmodule: Py_DECREF(structmodule) somewhere is needed.

You can check for refleak using: ./python -m test -R 3:3 test_buffer -m test_pybuffer_size_from_format

File: Objects/abstract.c [hunk 15]
```diff
@@ -495,6 +494,47 @@ _Py_add_one_to_index_C(int nd, Py_ssize_t *index, const Py_ssize_t *shape)
     }
 }
 
+Py_ssize_t
+PyBuffer_SizeFromFormat(const char *format)
+{
+    PyObject *structmodule = NULL;
+    PyObject *Struct = NULL;
```
Comment (vstinner) [inline Objects/abstract.c]: can you please rename it to "calcsize"? It's not a Struct class.

File: Modules/_testcapimodule.c [hunk 16]
```diff
@@ -3301,6 +3301,19 @@ getbuffer_with_null_view(PyObject* self, PyObject *obj)
     Py_RETURN_NONE;
 }
 
+/* test PyBuffer_SizeFromFormat() */
+static Py_ssize_t
```
Comment (vstinner) [inline Modules/_testcapimodule.c]: That's wrong: you must return PyObject*.

File: Modules/_testcapimodule.c [hunk 17]
```diff
@@ -3301,6 +3301,19 @@ getbuffer_with_null_view(PyObject* self, PyObject *obj)
...
+pybuffer_size_from_format(PyObject* self, PyObject *args)
+{
+    const char *format;
+
+    if (!PyArg_ParseTuple(args, "s:pybuffer_size_from_format",
+                          &format))
+        return -1;
+
+    return PyBuffer_SizeFromFormat(format);
```
Comment (vstinner) [inline Modules/_testcapimodule.c]: You can use for example:

```
    return PyLong_FromSsize_t(PyBuffer_SizeFromFormat(format));
```

File: Modules/_testcapimodule.c [hunk 18]
```diff
@@ -3301,6 +3301,19 @@ getbuffer_with_null_view(PyObject* self, PyObject *obj)
...
 }
 
+/* test PyBuffer_SizeFromFormat() */
+static Py_ssize_t
+pybuffer_size_from_format(PyObject* self, PyObject *args)
+{
+    const char *format;
+
+    if (!PyArg_ParseTuple(args, "s:pybuffer_size_from_format",
+                          &format))
```
Comment (vstinner) [inline Modules/_testcapimodule.c]: nitpick: please add {...}, I know that old code doesn't respect PEP 7, but new code should respect it.

**Comment by nanjekyejoannah** [comment]
have made the requested changes; please review again.

**Comment by nanjekyejoannah** [comment]
I have made the requested changes; please review again

File: Lib/test/test_buffer.py [hunk 19]
```diff
@@ -4412,6 +4417,13 @@ def test_issue_7385(self):
         x = ndarray([1,2,3], shape=[3], flags=ND_GETBUF_FAIL)
         self.assertRaises(BufferError, memoryview, x)
 
+    @support.cpython_only
```
Comment (vstinner) [inline Lib/test/test_buffer.py:4420]: @support.cpython_only should be removed: it's redundant with @unittest.skipUnless(_testcapi, 'requires _testcapi').

File: Misc/NEWS.d/next/Core and Builtins/2019-06-06-20-52-38.bpo-15913.5Sg5cv.rst [hunk 20]
```diff
@@ -0,0 +1 @@
+`PyBuffer_SizeFromFormat()` now added.
```
Comment (vstinner) [inline Misc/NEWS.d/next/Core and Builtins/2019-06-06-20-52-38.bpo-15913.5Sg5cv.rst]: Since it's documented, you can use the :c:func: markup to add a link:

```
:c:func:`PyBuffer_SizeFromFormat` now added.
```

File: Objects/abstract.c [hunk 21]
```diff
@@ -7,7 +7,6 @@
 #include "longintrepr.h"
 
 
-
```
Comment (vstinner) [inline Objects/abstract.c]: This change is unrelated, please revert it.

File: Doc/c-api/buffer.rst [hunk 22]
```diff
@@ -460,10 +460,12 @@ Buffer-related functions
    :c:func:`PyObject_GetBuffer`.
 
 
-.. c:function:: Py_ssize_t PyBuffer_SizeFromFormat(const char *)
+.. c:function:: Py_ssize_t PyBuffer_SizeFromFormat(const char *format)
 
-   Return the implied :c:data:`~Py_buffer.itemsize` from :c:data:`~Py_buffer.format`.
-   This function is not yet implemented.
+   Return the implied :c:data:`~Py_buffer.itemsize` from :c:data:`~Py_buffer.format`. on success, the
+   size of the return *format*. Otherwise, raise an exception and return -1 on error.
```
Comment (vstinner) [inline Doc/c-api/buffer.rst]: I don't understand "on success, the size of the return *format*." sentence, and it's redundant with the first sentence, no? I suggest to remove it, and just add:

On error, raise an exception and return -1.

**Review by vstinner — CHANGES_REQUESTED** [review]

File: Lib/test/test_buffer.py [hunk 23]
```diff
@@ -4412,6 +4417,12 @@ def test_issue_7385(self):
...
+    def test_pybuffer_size_from_format(self):
+        # basic tests
+        for format in ("i", "3s", "0i", " "):
+            self.assertEqual(_testcapi.pybuffer_size_from_format(format),
+                struct.calcsize(format.encode()))
```
Comment (vstinner) [inline Lib/test/test_buffer.py]: .encode() is useless and a little bit inconsistenct since format isn't encoded to call pybuffer_size_from_format().

Would you mind to remove the .encode() call (just pass format) but also indent the code so "struct.calcsize..." is aligned to "_testcapi.pybuffer..."?

**Review by vstinner — APPROVED** [review]
LGTM. But please make the last requested changes.

File: Objects/abstract.c [hunk 24]
```diff
@@ -495,6 +495,48 @@ _Py_add_one_to_index_C(int nd, Py_ssize_t *index, const Py_ssize_t *shape)
...
+{
+    PyObject *structmodule = NULL;
+    PyObject *calcsize = NULL;
+    PyObject *res = NULL;
+    PyObject *fmt = NULL;
+    Py_ssize_t itemsize = -1;
+
+    structmodule = PyImport_ImportModule("struct");
+    if (structmodule == NULL) {
+        goto done;
```
Comment (ZackerySpytz) [inline Objects/abstract.c]: If this `goto done;` is executed, a segfault will occur at `Py_DECREF(structmodule);`.

File: Modules/_testcapimodule.c [hunk 25]
```diff
@@ -3301,6 +3301,20 @@ getbuffer_with_null_view(PyObject* self, PyObject *obj)
...
+pybuffer_size_from_format(PyObject *self, PyObject *args)
+{
+    const char *format;
+
+    if (!PyArg_ParseTuple(args, "s:pybuffer_size_from_format",
+                          &format)) {
+        return NULL;
+    }
+
+    return PyLong_FromSsize_t(PyBuffer_SizeFromFormat(format));
```
Comment (ZackerySpytz) [inline Modules/_testcapimodule.c]: I don't believe `PyBuffer_SizeFromFormat()` should be inlined like this. Please check it for failure before the `PyLong_FromSsize_t()` call.

File: Objects/abstract.c [hunk 26]
```diff
@@ -495,6 +495,48 @@ _Py_add_one_to_index_C(int nd, Py_ssize_t *index, const Py_ssize_t *shape)
...
+    }
+
+    res = PyObject_CallFunctionObjArgs(calcsize, fmt, NULL);
+    if (res == NULL) {
+        goto done;
+    }
+
+    itemsize = PyLong_AsSsize_t(res);
+    if (itemsize < 0) {
```
Comment (ZackerySpytz) [inline Objects/abstract.c:528]: This check is not needed.

File: Objects/abstract.c (hunk 24 above)
Comment (nanjekyejoannah) [reply Objects/abstract.c]: How did you reproduce the segfault? I suspected a leak but verified none. I just ran one of the tests that use this function. using `./python -m test -R 3:3 test_buffer -m test_pybuffer_size_from_format`

Comment (vstinner) [reply Objects/abstract.c]: You can add "structmodule = NULL;" before the if to reproduce the crash.

File: Modules/_testcapimodule.c (hunk 25 above)
Comment (vstinner) [reply Modules/_testcapimodule.c]: Oh right, NULL should be returned if the result is -1.

File: Objects/abstract.c (hunk 26 above)
Comment (vstinner) [reply Objects/abstract.c:528]: calcsize() might return a integer which doesn't fit into a C Py_ssize_t. It's unlikely, but I prefer to keep the test.

File: Objects/abstract.c (hunk 24 above)
Comment (ZackerySpytz) [reply Objects/abstract.c]: The segfault would only occur if `PyImport_ImportModule()` failed (which is unlikely, but possible). It is enough to just use `return -1;` here instead of the `goto`.

File: Lib/test/test_buffer.py (hunk 23 above)
Comment (nanjekyejoannah) [reply Lib/test/test_buffer.py]: I have also seen that not encoding weirdly works and tests pass locally but will fail on the CI. So I tried to revert the .encode() on struct.calsize().

File: Objects/abstract.c (hunk 26 above)
Comment (ZackerySpytz) [reply Objects/abstract.c:528]: I know that. I meant that the explicit check is not needed because the control flow will fall through to the `done` label.

**Comment by vstinner** [comment]
Without .encode, the test failed with:

```
-----------------------------------
Traceback (most recent call last):
  File "/home/travis/build/python/cpython/Lib/test/test_buffer.py", line 4425, in test_pybuffer_size_from_format
    struct.calcsize(format))
BytesWarning: Comparison between bytes and string
```

You should be able to reproduce rhe issue using: ./python -bb -m tets test_buffer

The test should pass without .encode. You should investigate why it fails. struct.calcsize() accepts format as a Unicode string, no?

**Comment by nanjekyejoannah** [comment]
The test should pass without .encode. You should investigate why it fails. struct.calcsize() accepts format as a Unicode string, no?

I needed to change `fmt = PyBytes_FromString(format)` to `fmt = PyUnicode_FromString(format`); in order to do away with the comparison between bytes and string error.

The test passes now :

```
gitpod /workspace/cpython $ ./python -bb -m test test_buffer
Run tests sequentially
0:00:00 load avg: 0.69 [1/1] test_buffer

== Tests result: SUCCESS ==

1 test OK.

```

**Comment by nanjekyejoannah** [comment]
I have made the requested changes; please review again .

File: Misc/NEWS.d/next/Core and Builtins/2019-06-06-20-52-38.bpo-15913.5Sg5cv.rst [hunk 27]
```diff
@@ -0,0 +1 @@
+:c:func:`PyBuffer_SizeFromFormat()` now added.
```
Comment (vstinner) [inline Misc/NEWS.d/next/Core and Builtins/2019-06-06-20-52-38.bpo-15913.5Sg5cv.rst]: I suggest:

```Implement :c:func:`PyBuffer_SizeFromFormat()` function (previously documented but not implemented): call :func:`struct.calcsize`.```

File: Lib/test/test_buffer.py [hunk 28]
```diff
@@ -4412,6 +4417,12 @@ def test_issue_7385(self):
         x = ndarray([1,2,3], shape=[3], flags=ND_GETBUF_FAIL)
         self.assertRaises(BufferError, memoryview, x)
 
+    @support.cpython_only
+    def test_pybuffer_size_from_format(self):
+        nitems = randrange(1, 30)
```
Comment (vstinner) [inline Lib/test/test_buffer.py]: I would prefer to avoid using random in such test. iter_format() returns 91 items: it sounds overkill to me. IMHO tTesting 3 formats should but enough, like:

```
for fmt in ('', 'ii', '3s'):
```

File: Modules/_testcapimodule.c [hunk 29]
```diff
@@ -5087,6 +5107,7 @@ static PyMethodDef TestMethods[] = {
     {"test_pep3118_obsolete_write_locks", (PyCFunction)test_pep3118_obsolete_write_locks, METH_NOARGS},
 #endif
     {"getbuffer_with_null_view", getbuffer_with_null_view, METH_O},
+    {"pybuffer_size_from_format", (PyCFunction)pybuffer_size_from_format, METH_VARARGS},
```
Comment (vstinner) [inline Modules/_testcapimodule.c]: Maybe reuse the same name:
```suggestion
    {"PyBuffer_SizeFromFormat", test_PyBuffer_SizeFromFormat, METH_VARARGS},
```

Note: I don't think that (PyCFunction) cast is needed here.

**Review by vstinner — CHANGES_REQUESTED** [review]

File: Modules/_testcapimodule.c [hunk 30]
```diff
@@ -3301,6 +3301,26 @@ getbuffer_with_null_view(PyObject* self, PyObject *obj)
...
 }
 
+/* PyBuffer_SizeFromFormat() */
+static PyObject *
+test_PyBuffer_SizeFromFormat(PyObject *self, PyObject *args)
+{
+    const char *format;
+    Py_ssize_t result;
+
+    if (!PyArg_ParseTuple(args, "s:pybuffer_size_from_format",
```
Comment (vstinner) [inline Modules/_testcapimodule.c]: You forget to update the function name here.

File: Lib/test/test_buffer.py [hunk 31]
```diff
@@ -4412,6 +4417,11 @@ def test_issue_7385(self):
         x = ndarray([1,2,3], shape=[3], flags=ND_GETBUF_FAIL)
         self.assertRaises(BufferError, memoryview, x)
 
-
+    @support.cpython_only
+    def test_pybuffer_size_from_format(self):
+        # basic tests
+        for format in ("i", "3s", "0i", " "):
```
Comment (vstinner) [inline Lib/test/test_buffer.py]: " " <= my proposal was to test an empty format string

**Comment by nanjekyejoannah** [comment]
@vstinner I don't think this CI fail is now my business. Is it? i.e

`Windows PR Tests win64: "The job running on agent Azure Pipelines 7 has exceeded the maximum execution time of 60. For more information, see https://go.microsoft.com/fwlink/?linkid=2077134 "`

**Comment by vstinner** [comment]
I merged your PR @nanjekyejoannah, thanks!

> @vstinner I don't think this CI fail is now my business.

I don't know why Azure CI failed, but this CI is not mandatory yet.
